examples/bin/sample_client_server.rs

```diff
@@ -0,0 +1,216 @@
+                    }));
+                }
+            }
+            None
+        }
+
+        fn has_certs(&self) -> bool {
+            true
+        }
+    }
+
+    pub fn run_client(port: u16) -> Result<()> {
+        let hex_thumbprint = "a508d75eac3f646de59e406a5382ae2a40037d97";
+
+        let client_config =
+            ClientConfig::builder_with_provider(Arc::new(default_symcrypt_provider()))
+                .with_safe_default_protocol_versions()?
+                .dangerous()
+                .with_custom_certificate_verifier(Arc::new(
+                    Verifier::new().with_provider(Arc::new(default_symcrypt_provider())),
+                ))
+                .with_client_cert_resolver(Arc::new(ClientCertResolver(
+                    hex_thumbprint.to_string(),
+                )));
+
+        let mut connection =
+            ClientConnection::new(Arc::new(client_config), "rustls-server".try_into()?)?;
+        println!("start a new connection");
+        let mut client = TcpStream::connect(format!("localhost:{}", port))?;
+        let mut tls_stream = Stream::new(&mut connection, &mut client);
+        println!("write to tls stream");
+        tls_stream.write_all(b"ping")?;
+        println!("shut down socket");
+        tls_stream.sock.shutdown(Shutdown::Write)?;
+
+        let mut buf = [0u8; 4];
+        tls_stream.read_exact(&mut buf)?;
+        assert_eq!(&buf, b"pong");
+
+        tls_stream.sock.shutdown(Shutdown::Read)?;
+
+        Ok(())
+    }
+}
+
+mod server {
+    use std::{
+        io::{Read, Write},
+        net::{Shutdown, TcpListener, TcpStream},
+        sync::{mpsc::Sender, Arc},
+    };
+
+    use anyhow::Result;
+    use rustls::{
+        server::{ClientHello, ResolvesServerCert, ServerConfig, ServerConnection},
+        sign::CertifiedKey,
+        Stream,
+    };
+    use rustls_cng::{
+        signer::CngSigningKey,
+        store::{CertStore, CertStoreType},
+    };
+    use rustls_pki_types::CertificateDer;
+
+    #[derive(Debug)]
+    pub struct CachedServerCertResolver(Arc<CertifiedKey>);
+
+    impl CachedServerCertResolver {
+        pub fn new(chain: Vec<CertificateDer<'static>>, signing_key: CngSigningKey) -> Self {
+            let certified_key = CertifiedKey {
+                cert: chain,
+                key: Arc::new(signing_key),
+                ocsp: None,
+            };
+            CachedServerCertResolver(Arc::new(certified_key))
+        }
+    }
+
+    impl ResolvesServerCert for CachedServerCertResolver {
+        fn resolve(&self, _client_hello: ClientHello) -> Option<Arc<CertifiedKey>> {
+            Some(Arc::clone(&self.0))
+        }
+    }
+
+    fn get_chain(hex_thumbprint: &str) -> Result<(Vec<CertificateDer<'static>>, CngSigningKey)> {
+        let store = CertStore::open(CertStoreType::CurrentUser, "My")?;
+        let thumbprint = hex::decode(hex_thumbprint)?;
+
+        let context = store
+            .find_by_sha1(thumbprint)
```

Comment:
find_by_sha256

Reply:
not able to use it with current version 0.5.3.

Reply:
Update to depend on version of rustls-cng that supports.

Reply:
added comments.